Source/DafnyCore/Compilers/Compiler-java.cs

```diff
@@ -3679,28 +3716,73 @@ private class GenericArrayElementLvalue : ILvalue {
     protected override bool NeedsCastFromTypeParameter => true;
 
     protected override bool IsCoercionNecessary(Type/*?*/ from, Type/*?*/ to) {
-      if (from == null || to == null || !from.IsArrayType || !to.IsArrayType) {
-        return false;
+      if (UnicodeChars && ((IsObjectType(from) && to.IsCharType) || (from.IsCharType && IsObjectType(to)))) {
+        // Need to box from int to CodePoint, or unbox from CodePoint to int
+        return true;
       }
-      var dims = from.AsArrayType.Dims;
-      Contract.Assert(dims == to.AsArrayType.Dims);
-      if (dims > 1) {
+      
+      if (from == null || to == null) {
         return false;
       }
-      var udtFrom = (UserDefinedType)from.NormalizeExpand();
-      var udtTo = (UserDefinedType)to.NormalizeExpand();
-      if (!udtFrom.TypeArgs[0].IsTypeParameter || udtTo.TypeArgs[0].IsTypeParameter) {
-        return false;
+      if (from.IsArrayType && to.IsArrayType) {
+        var dims = from.AsArrayType.Dims;
+        Contract.Assert(dims == to.AsArrayType.Dims);
+        if (dims > 1) {
+          return false;
+        }
+
+        var udtFrom = (UserDefinedType)from.NormalizeExpand();
+        var udtTo = (UserDefinedType)to.NormalizeExpand();
+        return udtFrom.TypeArgs[0].IsTypeParameter && !udtTo.TypeArgs[0].IsTypeParameter;
```

Comment:
I don't know this part of the code well.  I think @fabiomadge understands the downcast stuff better than I do

Reply:
FYI it's not really related to the DowncastClone concept, it's a Java-specific problem involving having to manually box/unbox between an `int` and a `CodePoint` when interfacing with generics.